File: engine/scheduler.py
"""Inference scheduler: Continuous Batching + Chunked Prefill.

.. deprecated::
    The production scheduler is now the Rust `PyScheduler` in
    ``llm_serveforge_runtime``.  This Python implementation is retained
    as a reference and for standalone testing, but LLMEngine no longer
    uses it by default.


The scheduler is the runtime's central decision-maker. Each step:
  1. Removes finished requests from the running batch.
  2. Adds new requests from the waiting queue (priority-sorted min-heap).
  3. For prefill requests, applies Chunked Prefill to limit per-step tokens.
  4. Builds a Batch with input_ids, positions, and block_tables.

Key scheduling strategies:
    - FCFS (default): requests processed in arrival order.
    - Priority Queue: heapq-based, lower priority value = higher priority.
    - Chunked Prefill: long prompts split into chunk_size-token slices.
    - Hybrid: prefill and decode requests mixed in a single step.
    - Prefix Cache: RadixTree-based KV block reuse for shared prefixes.
"""

from __future__ import annotations

import heapq
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

import torch

from engine.batch import GenerationResult, Request, SamplingParams, SequenceGroup
from engine.block_manager import BlockManager

if TYPE_CHECKING:
    from cache.radix_cache import RadixCache


@dataclass(order=True)
class _QueueEntry:
    priority_value: int
    request: Request = field(compare=False)


class Scheduler:
    """Continuous Batching scheduler with Chunked Prefill support.

    Configuration:
        max_batch_size: Maximum number of concurrent requests in a batch.
        max_tokens_per_step: Total tokens processed per forward pass (prefill + decode).
        chunk_size: Maximum prefill tokens per request per step.

    Thread safety: the scheduler lock is held only during schedule().
    The caller must release it before calling forward() so new requests
    can be added to the waiting queue during GPU execution.
    """

    def __init__(
        self,
        max_batch_size: int = 32,
        max_tokens_per_step: int = 512,
        chunk_size: int = 256,
        radix_cache: RadixCache | None = None,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        if max_batch_size <= 0:
            raise ValueError(f"max_batch_size must be positive, got {max_batch_size}")
        if chunk_size <= 0:
            raise ValueError(f"chunk_size must be positive, got {chunk_size}")

        self.max_batch_size = max_batch_size
        self.max_tokens_per_step = max_tokens_per_step
        self.chunk_size = chunk_size
        self.radix_cache = radix_cache
        self._clock = clock

        self._waiting: list[_QueueEntry] = []
        self._running: list[Request] = []
        self._request_counter = 0

    def add_request(
        self,
        prompt_tokens: list[int],
        sampling_params: SamplingParams | None = None,
        priority: int = 0,
        request_id: str | None = None,
    ) -> str:
        if request_id is None:
            self._request_counter += 1
            request_id = f"req_{self._request_counter}"

        req = Request(
            request_id=request_id,
            prompt_tokens=list(prompt_tokens),
            sampling_params=sampling_params or SamplingParams(),
            priority=priority,
            arrival_time=self._clock(),
        )
        heapq.heappush(self._waiting, _QueueEntry(priority, req))
        return request_id

    def schedule(self, block_manager: BlockManager) -> SequenceGroup:
        self._reap_finished(block_manager)
        self._admit_requests(block_manager)

        if not self._running:
            return SequenceGroup()

        return self._build_sequence_group(block_manager)

    def _reap_finished(self, block_manager: BlockManager) -> None:
        finished = [r for r in self._running if r.is_finished]
        for r in finished:
            self._running.remove(r)
            if self.radix_cache is not None:
                all_tokens = r.prompt_tokens + r.output_tokens
                try:
                    blks = block_manager.get_blocks(r.request_id)
                    self.radix_cache.insert(all_tokens, blks)
                except KeyError:
                    pass
            block_manager.free(r.request_id)

    def _admit_requests(self, block_manager: BlockManager) -> None:
        while self._waiting and len(self._running) < self.max_batch_size:
            entry = heapq.heappop(self._waiting)
            req = entry.request
            req.state = "prefill"

            if self.radix_cache is not None:
                matched_blocks, matched_tokens = self.radix_cache.match_prefix(
                    req.prompt_tokens
                )
                if matched_tokens > 0:
                    req.prefill_pos = matched_tokens
                    block_manager.assign_cached_blocks(req.request_id, matched_blocks)
                    if matched_tokens >= len(req.prompt_tokens):
                        req.state = "decode"

            self._running.append(req)

    def _build_sequence_group(self, block_manager: BlockManager) -> SequenceGroup:
        batch_requests: list[Request] = []
        input_ids_list: list[torch.Tensor] = []
        positions_list: list[torch.Tensor] = []
        block_tables: dict[str, list[int]] = {}
        total_prefill_tokens = 0
        num_decode = 0

        for req in self._running:
            if req.state == "prefill":
                result = self._build_prefill_chunk(
                    req, block_manager, block_tables, total_prefill_tokens
                )
                if result is not None:
                    req_tensor, pos_tensor, n_tokens = result
                    batch_requests.append(req)
                    input_ids_list.append(req_tensor)
                    positions_list.append(pos_tensor)
                    total_prefill_tokens += n_tokens
                    if req.prefill_pos >= len(req.prompt_tokens):
                        req.state = "decode"
                continue

            if req.state == "decode":
                if total_prefill_tokens + num_decode + 1 > self.max_tokens_per_step:
                    continue

                num_decode += 1
                batch_requests.append(req)
                tok_tensor, pos_tensor, blocks = self._build_decode_step(req, block_manager)
                input_ids_list.append(tok_tensor)
                positions_list.append(pos_tensor)
                block_tables[req.request_id] = blocks

        if not batch_requests:
            return SequenceGroup()

        input_ids = torch.cat(input_ids_list) if input_ids_list else torch.tensor([], dtype=torch.long)
        positions = torch.cat(positions_list) if positions_list else torch.tensor([], dtype=torch.long)

        return SequenceGroup(
            requests=batch_requests,
            input_ids=input_ids,
            positions=positions,
            request_input_ids=input_ids_list,
            request_positions=positions_list,
            block_tables=block_tables,
        )

    def _build_prefill_chunk(
        self,
        req: Request,
        block_manager: BlockManager,
        block_tables: dict[str, list[int]],
        total_prefill_tokens: int,
    ) -> tuple[torch.Tensor, torch.Tensor, int] | None:
        remaining = req.tokens_remaining
        if remaining <= 0:
            req.state = "decode"
            return None

        prefill_budget = self.max_tokens_per_step - total_prefill_tokens
        if prefill_budget <= 0:
            return None

        n_tokens = min(remaining, self.chunk_size, prefill_budget)
        start_pos = req.prefill_pos
        end_pos = start_pos + n_tokens

        chunk_input_ids = req.prompt_tokens[start_pos:end_pos]
        req.prefill_pos = end_pos

        try:
            blocks = block_manager.get_blocks(req.request_id)
            covered = len(blocks) * block_manager.block_size
            if covered < len(req.prompt_tokens):
                extra = len(req.prompt_tokens) - covered
                n_extra = (
                    (extra + block_manager.block_size - 1)
                    // block_manager.block_size
                )
                for _ in range(n_extra):
                    bid = block_manager.free_blocks.pop()
                    block_manager.blocks[bid].ref_count = 1
                    block_manager.block_tables[req.request_id].append(bid)
        except KeyError:
            existing = block_manager.block_tables.get(req.request_id, [])
            existing_tokens = len(existing) * block_manager.block_size
            suffix_tokens = max(0, len(req.prompt_tokens) - existing_tokens)
            if suffix_tokens > 0:
                block_manager.allocate(req.request_id, suffix_tokens)
            else:
                block_manager.block_tables.setdefault(req.request_id, [])
            blocks = block_manager.get_blocks(req.request_id)

        block_tables[req.request_id] = blocks

        return (
            torch.tensor(chunk_input_ids, dtype=torch.long),
            torch.arange(start_pos, end_pos, dtype=torch.long),
            n_tokens,
        )

    def _build_decode_step(
        self, req: Request, block_manager: BlockManager
    ) -> tuple[torch.Tensor, torch.Tensor, list[int]]:
        if req.output_tokens:
            last_token = req.output_tokens[-1]
            pos = len(req.prompt_tokens) + len(req.output_tokens) - 1
        else:
            last_token = req.prompt_tokens[-1]
            pos = len(req.prompt_tokens) - 1

        try:
            blocks = block_manager.get_blocks(req.request_id)
        except KeyError:
            blocks = block_manager.allocate(req.request_id, pos + 1)

        return (
            torch.tensor([last_token], dtype=torch.long),
            torch.tensor([pos], dtype=torch.long),
            blocks,
        )

    def process_outputs(self, logits: torch.Tensor, batch: SequenceGroup) -> list[GenerationResult]:
        from engine.sampler import sample

        results: list[GenerationResult] = []
        if batch.is_empty or logits.numel() == 0:
            return results

        offset = 0
        for req in batch.requests:
            sp = req.sampling_params
            if req.state == "decode":
                req_logits = logits[offset : offset + 1]
                offset += 1
            elif req.state == "prefill":
                processed = req.num_processed_tokens
                if processed > 0 and processed <= logits.size(0):
                    req_logits = logits[processed - 1 : processed]
                else:
                    req_logits = logits[0:1]
            else:
                continue

            token_id = sample(
                req_logits,
                temperature=sp.temperature,
                top_p=sp.top_p,
                top_k=sp.top_k,
            )

            token_val = int(token_id.item())
            req.append_token(token_val)

            is_finished = False
            if len(req.output_tokens) >= sp.max_tokens:
                is_finished = True
            if token_val in sp.stop_token_ids:
                is_finished = True

            if is_finished:
                req.mark_finished()

            results.append(
                GenerationResult(
                    request_id=req.request_id,
                    new_tokens=[token_val],
                    is_finished=is_finished,
                )
            )

        return results

    @property
    def waiting_count(self) -> int:
        return len(self._waiting)

    @property
    def running_count(self) -> int:
        return len(self._running)

    @property
    def has_work(self) -> bool:
        return len(self._waiting) > 0 or len(self._running) > 0

File: rust/src/scheduler.rs
//! Continuous Batching scheduler with Chunked Prefill.
//!
//! Port of `engine/scheduler.py`.  The scheduler is the runtime's
//! central decision-maker.  Each `schedule()` call:
//!
//!   1. Reaps finished requests (frees blocks, records prefix cache hints).
//!   2. Admits new requests from the priority-sorted waiting queue.
//!   3. Builds prefill chunks respecting `chunk_size` and `max_tokens_per_step`.
//!   4. Builds decode steps (one token per request).
//!   5. Returns a `Batch` with per-request token/position metadata.
//!
//! Key strategies: FCFS (default), priority queue, chunked prefill,
//! hybrid prefill+decode mixing, prefix cache integration.

use std::cmp::Ordering;
use std::collections::BinaryHeap;

use crate::block_manager::BlockManager;
use crate::types::{Batch, PrefixCacheHit, Request, RequestState, ScheduledRequest};

// ── Priority queue entry for the waiting queue ──────────────

struct QueueEntry {
    priority: i32,
    request: Request,
}

impl Ord for QueueEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // Reverse: smaller priority = higher priority (min-heap)
        other.priority.cmp(&self.priority)
    }
}

impl PartialOrd for QueueEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl PartialEq for QueueEntry {
    fn eq(&self, other: &Self) -> bool {
        self.priority == other.priority
    }
}

impl Eq for QueueEntry {}

// ── Scheduler ───────────────────────────────────────────────

pub struct Scheduler {
    max_batch_size: usize,
    max_tokens_per_step: usize,
    chunk_size: usize,
    waiting: BinaryHeap<QueueEntry>,
    running: Vec<Request>,
    request_counter: usize,
}

impl Scheduler {
    pub fn new(
        max_batch_size: usize,
        max_tokens_per_step: usize,
        chunk_size: usize,
    ) -> Result<Self, String> {
        if max_batch_size == 0 {
            return Err("max_batch_size must be positive".into());
        }
        if chunk_size == 0 {
            return Err("chunk_size must be positive".into());
        }
        Ok(Self {
            max_batch_size,
            max_tokens_per_step,
            chunk_size,
            waiting: BinaryHeap::new(),
            running: Vec::new(),
            request_counter: 0,
        })
    }

    // ── Public API ──────────────────────────────────────────

    /// Add a request to the waiting queue.  Returns the assigned request ID.
    pub fn add_request(
        &mut self,
        prompt_tokens: Vec<u32>,
        priority: i32,
        arrival_time: f64,
        max_tokens: usize,
        stop_token_ids: Vec<u32>,
        request_id: Option<String>,
    ) -> String {
        let rid = request_id.unwrap_or_else(|| {
            self.request_counter += 1;
            format!("req_{}", self.request_counter)
        });
        let req = Request::new(
            rid.clone(),
            prompt_tokens,
            priority,
            arrival_time,
            max_tokens,
            stop_token_ids,
        );
        self.waiting.push(QueueEntry { priority, request: req });
        rid
    }

    /// Run one scheduling step.  Consumes prefix cache hints from Python.
    ///
    /// Returns a `Batch` with per-request metadata.  Python uses this
    /// to build `SequenceGroup` tensors.
    pub fn schedule(
        &mut self,
        block_manager: &mut BlockManager,
        cache_hits: &[PrefixCacheHit],
    ) -> Batch {
        let finished_request_ids: Vec<String> = self.running.iter()
            .filter(|r| r.is_finished())
            .map(|r| r.request_id.clone())
            .collect();

        // Reap finished requests
        self.running.retain(|r| !r.is_finished());
        for rid in &finished_request_ids {
            block_manager.free(rid);
        }

        // Admit new requests
        while let Some(entry) = self.waiting.pop() {
            if self.running.len() >= self.max_batch_size {
                self.waiting.push(entry);
                break;
            }
            let mut req = entry.request;
            req.state = RequestState::Prefill;

            // Apply prefix cache hits
            if let Some(hit) = cache_hits.iter().find(|h| h.request_id == req.request_id) {
                if hit.matched_tokens > 0 {
                    req.prefill_pos = hit.matched_tokens;
                    block_manager.assign_cached_blocks(&req.request_id, &hit.matched_blocks);
                    if hit.matched_tokens >= req.prompt_tokens.len() {
                        req.state = RequestState::Decode;
                    }
                }
            }

            self.running.push(req);
        }

        if self.running.is_empty() {
            return Batch::empty();
        }

        self.build_batch(block_manager)
    }

    // ── Internal: batch construction ────────────────────────

    fn build_batch(&mut self, block_manager: &mut BlockManager) -> Batch {
        let mut scheduled: Vec<ScheduledRequest> = Vec::new();
        let mut total_prefill_tokens: usize = 0;
        let mut num_decode: usize = 0;

        for i in 0..self.running.len() {
            let req = &mut self.running[i];
            match req.state {
                RequestState::Prefill => {
                    let remaining = req.tokens_remaining();
                    if remaining == 0 {
                        req.state = RequestState::Decode;
                        continue;
                    }

                    let prefill_budget = self.max_tokens_per_step
                        .saturating_sub(total_prefill_tokens);
                    if prefill_budget == 0 {
                        continue;
                    }

                    let n_tokens = remaining.min(self.chunk_size).min(prefill_budget);
                    let start_pos = req.prefill_pos;
                    let end_pos = start_pos + n_tokens;

                    let chunk_ids: Vec<u32> = req.prompt_tokens[start_pos..end_pos].to_vec();
                    let positions: Vec<u32> = (start_pos as u32..end_pos as u32).collect();
                    req.prefill_pos = end_pos;

                    // Ensure blocks exist — skip request on OOM so it retries next step
                    if block_manager.block_tables.contains_key(&req.request_id) {
                        if block_manager.ensure_blocks(&req.request_id, req.prompt_tokens.len()).is_err() {
                            continue; // OOM, retry next step
                        }
                    } else if block_manager.allocate(&req.request_id, req.prompt_tokens.len()).is_err() {
                        continue; // OOM, retry next step
                    }

                    let blocks = match block_manager.get_blocks(&req.request_id) {
                        Ok(blks) => blks.to_vec(),
                        Err(_) => {
                            // Block table missing after successful allocation — logic error.
                            // Skip request to avoid passing empty block_table to Python.
                            continue;
                        }
                    };

                    if req.prefill_pos >= req.prompt_tokens.len() {
                        req.state = RequestState::Decode;
                    }

                    scheduled.push(ScheduledRequest {
                        request_id: req.request_id.clone(),
                        input_ids: chunk_ids,
                        positions,
                        state: RequestState::Prefill,
                        block_table: blocks,
                        n_tokens,
                    });

                    total_prefill_tokens += n_tokens;
                }

                RequestState::Decode => {
                    if total_prefill_tokens + num_decode + 1 > self.max_tokens_per_step {
                        continue;
                    }

                    num_decode += 1;

                    let (last_token, pos) = if let Some(last) = req.output_tokens.last() {
                        (*last, (req.prompt_tokens.len() + req.output_tokens.len() - 1) as u32)
                    } else {
                        let last = req.prompt_tokens[req.prompt_tokens.len() - 1];
                        (last, (req.prompt_tokens.len() - 1) as u32)
                    };

                    let blocks = match block_manager.get_blocks(&req.request_id) {
                        Ok(blks) => blks.to_vec(),
                        Err(_) => continue,
                    };

                    scheduled.push(ScheduledRequest {
                        request_id: req.request_id.clone(),
                        input_ids: vec![last_token],
                        positions: vec![pos],
                        state: RequestState::Decode,
                        block_table: blocks,
                        n_tokens: 1,
                    });
                }

                _ => {}
            }
        }

        let total_tokens: usize = scheduled.iter().map(|s| s.n_tokens).sum();
        Batch {
            requests: scheduled,
            total_tokens,
        }
    }

    // ── Query helpers ───────────────────────────────────────

    pub fn waiting_count(&self) -> usize {
        self.waiting.len()
    }

    pub fn running_count(&self) -> usize {
        self.running.len()
    }

    pub fn has_work(&self) -> bool {
        !self.waiting.is_empty() || !self.running.is_empty()
    }

    /// Get a reference to a running request by ID.
    pub fn running_request(&self, request_id: &str) -> Option<&Request> {
        self.running.iter().find(|r| r.request_id == request_id)
    }

    /// Get a finished request by ID (removes it from running list).
    pub fn get_finished_request(&mut self, request_id: &str) -> Option<Request> {
        let idx = self.running.iter().position(|r| r.request_id == request_id)?;
        Some(self.running.remove(idx))
    }

    /// Record output tokens for a request and check for termination.
    ///
    /// Returns `true` if the request should be marked finished.
    pub fn record_output(&mut self, request_id: &str, token_id: u32) -> bool {
        let Some(req) = self.running.iter_mut().find(|r| r.request_id == request_id) else {
            return false;
        };
        req.append_token(token_id);

        if req.output_tokens.len() >= req.max_tokens {
            req.mark_finished();
            return true;
        }
        if req.stop_token_ids.contains(&token_id) {
            req.mark_finished();
            return true;
        }
        false
    }
}

// ── Tests ───────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn make_block_manager() -> BlockManager {
        BlockManager::new(1000, 16).unwrap()
    }

    #[test]
    fn test_scheduler_creation() {
        let s = Scheduler::new(32, 512, 256).unwrap();
        assert_eq!(s.max_batch_size, 32);
        assert_eq!(s.chunk_size, 256);
        assert!(s.waiting.is_empty());
        assert!(s.running.is_empty());
    }

    #[test]
    fn test_invalid_params() {
        assert!(Scheduler::new(0, 512, 256).is_err());
        assert!(Scheduler::new(32, 512, 0).is_err());
    }

    #[test]
    fn test_add_request_returns_id() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let rid = s.add_request(vec![1, 2, 3], 0, 0.0, 256, vec![], None);
        assert!(rid.starts_with("req_"));
        assert_eq!(s.waiting_count(), 1);
    }

    #[test]
    fn test_add_request_with_custom_id() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let rid = s.add_request(vec![1, 2], 0, 0.0, 256, vec![], Some("my_id".into()));
        assert_eq!(rid, "my_id");
    }

    #[test]
    fn test_empty_schedule_returns_empty_batch() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let mut bm = make_block_manager();
        let batch = s.schedule(&mut bm, &[]);
        assert!(batch.is_empty());
    }

    #[test]
    fn test_single_request_prefill() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let mut bm = make_block_manager();
        s.add_request(vec![1, 2, 3, 4, 5], 0, 0.0, 256, vec![], None);
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests.len(), 1);
        assert_eq!(batch.requests[0].state, RequestState::Prefill);
        assert_eq!(batch.requests[0].input_ids, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_request_transitions_to_decode() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let mut bm = make_block_manager();
        s.add_request(vec![1, 2, 3], 0, 0.0, 256, vec![], None);

        // First schedule: prefill all 3 tokens
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests.len(), 1);
        assert_eq!(batch.requests[0].state, RequestState::Prefill);
        assert_eq!(batch.requests[0].n_tokens, 3);

        // Second schedule: now in decode
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests.len(), 1);
        assert_eq!(batch.requests[0].state, RequestState::Decode);
    }

    #[test]
    fn test_chunked_prefill_splits_long_prompt() {
        let mut s = Scheduler::new(32, 512, 4).unwrap(); // chunk_size=4
        let mut bm = make_block_manager();
        let prompt: Vec<u32> = (0..10).collect();
        s.add_request(prompt, 0, 0.0, 256, vec![], None);

        // First step: 4 tokens (chunk_size)
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests.len(), 1);
        assert_eq!(batch.requests[0].state, RequestState::Prefill);
        assert_eq!(batch.requests[0].n_tokens, 4);

        // Second step: next 4 tokens
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests[0].n_tokens, 4);

        // Third step: last 2 tokens
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests[0].n_tokens, 2); // remaining

        // Fourth step: decode
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests[0].state, RequestState::Decode);
    }

    #[test]
    fn test_priority_queue_order() {
        let mut s = Scheduler::new(1, 512, 256).unwrap(); // batch_size=1 to test ordering
        let mut bm = make_block_manager();

        s.add_request(vec![7], 0, 0.0, 256, vec![], None);
        s.add_request(vec![2], 5, 0.0, 256, vec![], None);
        s.add_request(vec![3], 10, 0.0, 256, vec![], None);

        // First admitted should be priority 0 (lowest value = highest priority)
        let batch = s.schedule(&mut bm, &[]);
        assert_eq!(batch.requests.len(), 1);
        assert_eq!(batch.requests[0].request_id, "req_1");
    }

    #[test]
    fn test_finished_request_reaped() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        let mut bm = make_block_manager();

        let rid = s.add_request(vec![1], 0, 0.0, 256, vec![], None);
        let _ = s.schedule(&mut bm, &[]); // prefill
        let _ = s.schedule(&mut bm, &[]); // decode starts

        // Mark as finished
        s.record_output(&rid, 42); // not a stop token, check max_tokens
        // Fast-forward past max_tokens
        for _ in 0..255 {
            if let Some(req) = s.running.iter_mut().find(|r| r.request_id == rid) {
                req.append_token(42);
            }
        }
        s.record_output(&rid, 99); // Should hit max_tokens=256

        // Next schedule should reap
        let batch = s.schedule(&mut bm, &[]);
        assert!(batch.is_empty());
    }

    #[test]
    fn test_has_work() {
        let mut s = Scheduler::new(32, 512, 256).unwrap();
        assert!(!s.has_work());
        s.add_request(vec![1], 0, 0.0, 256, vec![], None);
        assert!(s.has_work());
    }
}
